src/views/shared/private-route/private-route.view.tsx

```diff
@@ -1,20 +1,63 @@
-import { FC, PropsWithChildren } from "react";
+import { FC, PropsWithChildren, useEffect, useState } from "react";
 import { Navigate, useLocation } from "react-router-dom";
+import { useEnvContext } from "src/contexts/env.context";
 
 import { useProvidersContext } from "src/contexts/providers.context";
 import routes from "src/routes";
+import { isAsyncTaskDataAvailable } from "src/utils/types";
 
 const PrivateRoute: FC<PropsWithChildren> = ({ children }) => {
-  const { connectedProvider } = useProvidersContext();
+  const env = useEnvContext();
+  const {
+    connectedProvider,
+    account,
+    getMetaMaskProvider,
+    connectMetaMaskProvider,
+    silentlyGetMetaMaskConnectedAccounts,
+  } = useProvidersContext();
   const { pathname, search } = useLocation();
+  const [nextRoute, setNextRoute] = useState<"currentRoute" | "login">();
 
-  if (!connectedProvider) {
-    return (
-      <Navigate to={routes.login.path} replace state={{ redirectUrl: `${pathname}${search}` }} />
-    );
-  }
+  useEffect(() => {
+    if (connectedProvider && isAsyncTaskDataAvailable(account)) {
+      setNextRoute("currentRoute");
+    } else if (!connectedProvider && env) {
+      const web3Provider = getMetaMaskProvider();
+
+      if (web3Provider) {
+        void silentlyGetMetaMaskConnectedAccounts({ web3Provider }).then((accounts) => {
+          const account = accounts ? accounts[0] : undefined;
 
-  return <>{children}</>;
+          if (!account) {
+            setNextRoute("login");
+          } else {
+            return connectMetaMaskProvider({ env, web3Provider, account });
+          }
+        });
+      }
+    }
+  }, [
+    account,
+    connectedProvider,
+    env,
+    connectMetaMaskProvider,
+    getMetaMaskProvider,
+    silentlyGetMetaMaskConnectedAccounts,
+  ]);
+
+  switch (nextRoute) {
+    case "currentRoute": {
+      return <>{children}</>;
+    }
+    case "login": {
+      return (
+        <Navigate to={routes.login.path} replace state={{ redirectUrl: `${pathname}${search}` }} />
+      );
+    }
+    default: {
+      return null;
+    }
+  }
```

Comment:
Similar to the [comment in the login view](https://github.com/0xPolygonHermez/zkevm-bridge-ui/pull/193/files#r995531151), here we're overcomplicating things by having an extra state `nextRoute` and a `useEffect` instantiating the provider (this should ideally only happen in the provider context), and calling `silentlyGetMetaMaskConnectedAccounts` and `connectMetaMaskProvider`.

Before, using the `isProviderConnecting` flag this was as simple as:

```
  if (isProviderConnecting) {
    return null;
  } else if (!connectedProvider) {
    return (
      <Navigate to={routes.login.path} replace state={{ redirectUrl: `${pathname}${search}` }} />
    );
  } else {
    return <>{children}</>;
  }
```

I'm not convinced by all these changes to be honest.

While I agree with your initial comment regarding the `isProviderConnecting` flag in the provider being a bit hard to follow and bug-prone, it has the benefit of encapsulating all the provider-related logic (instantiating, connecting, etc) in the Provider module. This alternative is spreading the logic to the views and making things more complex outside the Provider module.

I think we need to keep thinking about this to find a better solution.

Random thought, what if, instead of adding the `isProviderConnecting` flag, we convert the provider to an async task (or a similar interface) so we can have it in "loading"/"logging-in" and the views return these nulls while this is its state?

Maybe we can have a meeting to keep discussing all this. Luckily, we're not in a hurry for this one.